packages/iframe-stamper/src/index.ts

```diff
@@ -145,36 +155,58 @@ export class IframeStamper {
 
     // This is populated once the iframe is ready. Call `.init()` to kick off DOM insertion!
     this.iframePublicKey = null;
+
+    /**
+     * The MessageChannel API is used to establish secure communication between the parent page and the iframe.
+     * See https://developer.mozilla.org/en-US/docs/Web/API/MessageChannel
+     */
+    this.messageChannel = new MessageChannel();
   }
 
   /**
    * Inserts the iframe on the page and returns a promise resolving to the iframe's public key
    */
   async init(): Promise<string> {
     this.container.appendChild(this.iframe);
-    return new Promise((resolve, _reject) => {
-      window.addEventListener(
-        "message",
-        (event) => {
-          if (event.origin !== this.iframeOrigin) {
-            // There might be other things going on in the window, for example: react dev tools, other extensions, etc.
-            // Instead of erroring out
-            return;
-          }
-          if (event.data?.type === IframeEventType.PublicKeyReady) {
-            this.iframePublicKey = event.data["value"];
-            resolve(event.data["value"]);
-          }
-        },
-        false
+    /**
+     * Once the iframe is loaded, we send a message to the iframe to hand over the
+     * MessageChannel's second port, port2, and establish the secure communication channel.
+     * The iframe will use this port to send messages back to the parent page.
+     * See https://developer.mozilla.org/en-US/docs/Web/API/MessagePort/postMessage#transfer
+     */
+    this.iframe.addEventListener("load", () => {
+      if (
+        !this.iframe.contentWindow ||
+        !this.iframe.contentWindow.postMessage
+      ) {
+        throw new Error(
+          "contentWindow or contentWindow.postMessage does not exist"
+        );
+      }
+
+      this.iframe.contentWindow.postMessage(
+        { type: IframeEventType.TurnkeyInitMessageChannel },
+        this.iframeOrigin,
+        [this.messageChannel.port2]
       );
     });
+
+    return new Promise((resolve, _reject) => {
+      this.messageChannel.port1.onmessage = (event) => {
```

Comment:
non-blocking: a quick note on the importance of the port1 (might be trivial given you've explained port2) could be helpful